// @polymarket/order-utils/dist/exchange.order.const.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ORDER_STRUCTURE = exports.EIP712_DOMAIN = exports.EXCHANGE_ABI = exports.CALL_RESULTS_PREFIX = exports.ZERO_ADDRESS = exports.ZX = exports.PROTOCOL_VERSION = exports.PROTOCOL_NAME = void 0;
const tslib_1 = require("tslib");
const Exchange_json_1 = tslib_1.__importDefault(require("./abi/Exchange.json"));
// Contract's constants.
exports.PROTOCOL_NAME = 'Polymarket CTF Exchange';
exports.PROTOCOL_VERSION = '1';
exports.ZX = '0x';
exports.ZERO_ADDRESS = '0x0000000000000000000000000000000000000000';
exports.CALL_RESULTS_PREFIX = 'CALL_RESULTS_';
// ABIs
exports.EXCHANGE_ABI = Exchange_json_1.default;
// EIP712 Objects
exports.EIP712_DOMAIN = [
    { name: 'name', type: 'string' },
    { name: 'version', type: 'string' },
    { name: 'chainId', type: 'uint256' },
    { name: 'verifyingContract', type: 'address' },
];
exports.ORDER_STRUCTURE = [
    { name: 'salt', type: 'uint256' },
    { name: 'maker', type: 'address' },
    { name: 'signer', type: 'address' },
    { name: 'taker', type: 'address' },
    { name: 'tokenId', type: 'uint256' },
    { name: 'makerAmount', type: 'uint256' },
    { name: 'takerAmount', type: 'uint256' },
    { name: 'expiration', type: 'uint256' },
    { name: 'nonce', type: 'uint256' },
    { name: 'feeRateBps', type: 'uint256' },
    { name: 'side', type: 'uint8' },
    { name: 'signatureType', type: 'uint8' },
];
