// @polymarket/clob-client/dist/http-helpers/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseDropNotificationParams = exports.parseOrdersScoringParams = exports.del = exports.get = exports.post = exports.request = exports.PUT = exports.DELETE = exports.POST = exports.GET = void 0;
const tslib_1 = require("tslib");
/* eslint-disable max-depth */
const axios_1 = tslib_1.__importDefault(require("axios"));
const browser_or_node_1 = require("browser-or-node");
exports.GET = "GET";
exports.POST = "POST";
exports.DELETE = "DELETE";
exports.PUT = "PUT";
const overloadHeaders = (method, headers) => {
    if (browser_or_node_1.isBrowser) {
        return;
    }
    if (!headers || typeof headers === undefined) {
        headers = {};
    }
    if (headers) {
        headers["User-Agent"] = `@polymarket/clob-client`;
        headers["Accept"] = "*/*";
        headers["Connection"] = "keep-alive";
        headers["Content-Type"] = "application/json";
        if (method === exports.GET) {
            headers["Accept-Encoding"] = "gzip";
        }
    }
};
const request = (endpoint, method, headers, data, params) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    overloadHeaders(method, headers);
    return yield (0, axios_1.default)({ method, url: endpoint, headers, data, params });
});
exports.request = request;
const post = (endpoint, options) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    try {
        const resp = yield (0, exports.request)(endpoint, exports.POST, options === null || options === void 0 ? void 0 : options.headers, options === null || options === void 0 ? void 0 : options.data, options === null || options === void 0 ? void 0 : options.params);
        return resp.data;
    }
    catch (err) {
        return errorHandling(err);
    }
});
exports.post = post;
const get = (endpoint, options) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    try {
        const resp = yield (0, exports.request)(endpoint, exports.GET, options === null || options === void 0 ? void 0 : options.headers, options === null || options === void 0 ? void 0 : options.data, options === null || options === void 0 ? void 0 : options.params);
        return resp.data;
    }
    catch (err) {
        return errorHandling(err);
    }
});
exports.get = get;
const del = (endpoint, options) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    try {
        const resp = yield (0, exports.request)(endpoint, exports.DELETE, options === null || options === void 0 ? void 0 : options.headers, options === null || options === void 0 ? void 0 : options.data, options === null || options === void 0 ? void 0 : options.params);
        return resp.data;
    }
    catch (err) {
        return errorHandling(err);
    }
});
exports.del = del;
const errorHandling = (err) => {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
    if (axios_1.default.isAxiosError(err)) {
        if (err.response) {
            console.error("[CLOB Client] request error", JSON.stringify({
                status: (_a = err.response) === null || _a === void 0 ? void 0 : _a.status,
                statusText: (_b = err.response) === null || _b === void 0 ? void 0 : _b.statusText,
                data: (_c = err.response) === null || _c === void 0 ? void 0 : _c.data,
                config: (_d = err.response) === null || _d === void 0 ? void 0 : _d.config,
            }));
            if ((_e = err.response) === null || _e === void 0 ? void 0 : _e.data) {
                if (typeof ((_f = err.response) === null || _f === void 0 ? void 0 : _f.data) === "string" ||
                    ((_g = err.response) === null || _g === void 0 ? void 0 : _g.data) instanceof String) {
                    return { error: (_h = err.response) === null || _h === void 0 ? void 0 : _h.data };
                }
                if (!Object.prototype.hasOwnProperty.call((_j = err.response) === null || _j === void 0 ? void 0 : _j.data, "error")) {
                    return { error: (_k = err.response) === null || _k === void 0 ? void 0 : _k.data };
                }
                // in this case the field 'error' is included
                return (_l = err.response) === null || _l === void 0 ? void 0 : _l.data;
            }
        }
        if (err.message) {
            console.error("[CLOB Client] request error", JSON.stringify({
                error: err.message,
            }));
            return { error: err.message };
        }
    }
    console.error("[CLOB Client] request error", err);
    return { error: err };
};
const parseOrdersScoringParams = (orderScoringParams) => {
    const params = {};
    if (orderScoringParams !== undefined) {
        if (orderScoringParams.orderIds !== undefined) {
            params["order_ids"] = orderScoringParams === null || orderScoringParams === void 0 ? void 0 : orderScoringParams.orderIds.join(",");
        }
    }
    return params;
};
exports.parseOrdersScoringParams = parseOrdersScoringParams;
const parseDropNotificationParams = (dropNotificationParams) => {
    const params = {};
    if (dropNotificationParams !== undefined) {
        if (dropNotificationParams.ids !== undefined) {
            params["ids"] = dropNotificationParams === null || dropNotificationParams === void 0 ? void 0 : dropNotificationParams.ids.join(",");
        }
    }
    return params;
};
exports.parseDropNotificationParams = parseDropNotificationParams;
